kernel/src/log_replay.rs

```diff
@@ -190,3 +194,100 @@ impl<'seen> FileActionDeduplicator<'seen> {
+///
+/// Log replay processors scan transaction logs in **reverse chronological order** (newest to oldest),
+/// filtering and transforming action batches into specialized output types. These processors:
+///
+/// - **Track and deduplicate file actions** to ensure only the latest relevant changes are kept.
+/// - **Maintain selection vectors** to indicate which actions in each batch should be included.
+/// - **Apply custom filtering logic** based on the processor’s purpose (e.g., checkpointing, scanning).
+///
+/// Implementations:
+/// - `ScanLogReplayProcessor`: Used for table scans, this processor filters and selects deduplicated  
+///   `Add` actions from log batches to reconstruct the view of the table at a specific point in time.
+///   Note that scans do not expose `Remove` actions.
+/// - `V1CheckpointLogReplayProcessor`(WIP): Will be responsible for processing log batches to construct  
+///   V1 spec checkpoint files. Unlike scans, checkpoint processing includes additional actions,  
+///   such as `Remove`, `Metadata`, and `Protocol`, required to fully reconstruct table state.
+///
+/// The `Output` type must implement [`HasSelectionVector`] to enable filtering of batches
+/// with no selected rows.
+///
+/// TODO: Refactor the Change Data Feed (CDF) processor to use this trait.
+pub(crate) trait LogReplayProcessor: Sized {
+    /// The type of results produced by this processor must implement the
+    /// `HasSelectionVector` trait to allow filtering out batches with no selected rows.
+    type Output: HasSelectionVector;
+
+    /// Processes a batch of actions and returns the filtered results.
+    ///
+    /// # Arguments
+    /// - `actions_batch` - A boxed [`EngineData`] instance representing a batch of actions.
+    /// - `is_log_batch` - `true` if the batch originates from a commit log, `false` if from a checkpoint.
+    ///
+    /// Returns a [`DeltaResult`] containing the processor’s output, which includes only selected actions.
+    ///
+    /// Note: Since log replay is stateful, processing may update internal processor state (e.g., deduplication sets).
+    fn process_actions_batch(
+        &mut self,
+        actions_batch: Box<dyn EngineData>,
+        is_log_batch: bool,
+    ) -> DeltaResult<Self::Output>;
+
+    /// Applies the processor to an actions iterator and filters out empty results.
+    ///
+    /// # Arguments
+    /// * `action_iter` - Iterator of action batches and their source flags
+    ///
+    /// Returns an iterator that yields the Output type of the processor.
+    fn process_actions_iter(
+        mut self,
+        action_iter: impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>>,
+    ) -> impl Iterator<Item = DeltaResult<Self::Output>> {
+        action_iter
+            .map(move |action_res| {
+                let (batch, is_log_batch) = action_res?;
+                self.process_actions_batch(batch, is_log_batch)
+            })
```

Comment:
nit: Is it worth passing the pair directly to `process_actions_batch`?
```suggestion
            .map(move |action_res| self.process_actions_batch(action_res?))
```
Especially if we add an `ActionBatch` struct or type alias for `(Box<dyn EngineData>, bool)`, which would simplify a half dozen existing use sites?

Reply:
ahh huge +1 this is a great idea - i've noticed we have many places where `EngineData` is coupled with some metadata (this bool, or selection vector, etc.) and I think it vastly helps understandability to have these be types themselves (ActionBatch, FilteredBatch, etc.)

Reply:
Thanks! Totally agree, the extra structure makes the data flow way easier to follow. I've filed a follow-up issue to introduce an `ActionBatch` type and update the call sites #806